File: app/src/main/java/com/qzb/scoketlink/wifip2p/WiFiDirectBroadcastReceiver.java
package com.qzb.scoketlink.wifip2p;

import android.Manifest;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.wifi.p2p.WifiP2pManager;
import android.util.Log;

import androidx.core.app.ActivityCompat;

/**
 * A BroadcastReceiver that notifies of important Wi-Fi p2p events.
 */
public class WiFiDirectBroadcastReceiver extends BroadcastReceiver {

    private WifiP2pManager mManager;
    private WifiP2pManager.Channel mChannel;
    private WifiP2pManager.PeerListListener peerListListener;
    private Context context;

    public WiFiDirectBroadcastReceiver(Context context, WifiP2pManager manager, WifiP2pManager.Channel channel, WifiP2pManager.PeerListListener peerListListener) {
        super();
        this.context = context;
        this.mManager = manager;
        this.mChannel = channel;
        this.peerListListener = peerListListener;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        String action = intent.getAction();
        Log.d("wifip2p", "action:" + action);

        if (WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION.equals(action)) {
            // 检查 WLAN P2P 是否开启并受支持
            int state = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE, -1);
            if (state == WifiP2pManager.WIFI_P2P_STATE_ENABLED) {
                // Wifi P2P is enabled
                Log.d("wifip2p", "p2p enable:" + true);
            } else {
                // Wi-Fi P2P is not enabled
                Log.d("wifip2p", "p2p enable:" + false);
            }
        } else if (WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION.equals(action)) {
            Log.d("wifip2p", "requestPeers111");
            // Call WifiP2pManager.requestPeers() to get a list of current peers
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                return;
            }
            Log.d("wifip2p", "requestPeers");
            mManager.requestPeers(mChannel, peerListListener);
        } else if (WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION.equals(action)) {
            // Respond to new connection or disconnections
        } else if (WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION.equals(action)) {
            // Respond to this device's wifi state changing
        }
    }
}

File: smaple/src/main/java/com/qzb/smaple/wifip2p/WiFiDirectBroadcastReceiver.kt
package com.qzb.smaple.wifip2p

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.wifi.p2p.WifiP2pManager

@SuppressLint("MissingPermission")
class WiFiDirectBroadcastReceiver(
    private val listener: P2PListener
) : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
        when (intent?.action) {
            WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION -> listener.onP2PStateChanged(intent)
            WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION -> listener.onP2PPeersChanged(intent)
            WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION ->listener.onP2PConnectionChanged(intent)
            WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION -> listener.onP2PThisDeviceChanged(intent)
        }
    }

    public interface P2PListener {
        fun onP2PStateChanged(intent: Intent)
        fun onP2PPeersChanged(intent: Intent)
        fun onP2PConnectionChanged(intent: Intent)
        fun onP2PThisDeviceChanged(intent: Intent)
    }
}